src/Init/Data/Repr.lean

```diff
@@ -51,6 +51,12 @@ instance [Repr α] : Repr (id α) :=
 instance [Repr α] : Repr (Id α) :=
   inferInstanceAs (Repr α)
 
+/-
+This instance allows us to use `Empty` as a type parameter without causing automatic
+derivations of `Repr` to fail.
+-/
+deriving instance Repr for Empty
```

Comment:
We may as well write the instance explicitly rather than rely on the deriving machinery.
```suggestion
instance : Repr Empty where
  reprPrec := nofun
```
I'm not sure we need the comment justifying the instance, but I suppose it doesn't hurt.

Reply:
Will fix.

Just for my understanding, when would I chose one approach over the other?  It seems `deriving` is used lower down in the same file.  Also, strictly speaking, I wouldn't have known how to write that instance because I'm not familiar with `nofun` (and `nomatch` which comes up when I hover).

I found the definition of `nofun` in `Match.lean` (the elaborator! ;) ) but I can't quite read that code yet.  Is there a short summary/explanation of the semantics you may be able to point me at, please?

Reply:
> I'm not sure we need the comment justifying the instance

Please let me know if you have a preference.  I opted to be more explicit but that's just because I wanted to "do the right thing".  Comments which are obvious to experienced readers are a nuisance, but I'm not the right audience to ask whether the need for `Repr` on `Empty` is obvious or not.

Reply:
Let's leave the comment, it's helpful knowing why we'd have a "useless" instance.

> when would I chose one approach over the other?

You generally want to use `deriving`, but I just figured that this one is so simple that we may as well give the definition. The `nofun` syntax is `fun | patt1 => ... | patt2 => ... | ...` syntax but with no patterns, which itself is short for `fun x => match x with | patt1 => ... | patt2 => ... | ...` syntax. You can also write `fun x => nomatch x` instead of `nofun`.

Reply:
Thanks for the thorough explanation.

Reply:
Side note: I've updated the PR description to clarify what the change is for. The terminology is "typeclass inference" or "instance synthesis" instead of "deriving" for what causes the issue in the `#eval` command. Deriving is what occurs at the `deriving` clause.

Reply:
Noted with thanks.